src/Features/CSharp/Portable/QuickInfo/CSharpDiagnosticAnalyzerQuickInfoProvider.cs

```diff
@@ -0,0 +1,182 @@
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.QuickInfo;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.QuickInfo
+{
+    [ExportQuickInfoProvider(QuickInfoProviderNames.DiagnosticAnalyzer, LanguageNames.CSharp), Shared]
+    [ExtensionOrder(Before = QuickInfoProviderNames.Semantic)]
+    internal class CSharpDiagnosticAnalyzerQuickInfoProvider : CommonQuickInfoProvider
+    {
+        private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpDiagnosticAnalyzerQuickInfoProvider(IDiagnosticAnalyzerService diagnosticAnalyzerService)
+        {
+            _diagnosticAnalyzerService = diagnosticAnalyzerService;
+        }
+
+        protected override async Task<QuickInfoItem?> BuildQuickInfoAsync(
+            Document document,
+            SyntaxToken token,
+            CancellationToken cancellationToken)
+        {
+            return GetQuickinfoForPragmaWarning(document, token) ??
+                (await GetQuickInfoForSuppressMessageAttributeAsync(document, token, cancellationToken).ConfigureAwait(false));
+        }
+
+        private QuickInfoItem? GetQuickinfoForPragmaWarning(Document document, SyntaxToken token)
+        {
+            var errorCode = token.Parent switch
+            {
+                PragmaWarningDirectiveTriviaSyntax directive
+                    => token.IsKind(SyntaxKind.EndOfDirectiveToken)
+                        ? directive.ErrorCodes.LastOrDefault() as IdentifierNameSyntax
+                        : directive.ErrorCodes.FirstOrDefault() as IdentifierNameSyntax,
+                IdentifierNameSyntax { Parent: PragmaWarningDirectiveTriviaSyntax _ } identifier
+                    => identifier,
+                _ => null,
+            };
+
+            if (errorCode != null)
+            {
+                return GetQuickInfoFromSupportedDiagnosticsOfProjectAnalyzers(document, errorCode.Identifier.ValueText, errorCode.Span);
+            }
+
+            return null;
+        }
+
+        private async Task<QuickInfoItem?> GetQuickInfoForSuppressMessageAttributeAsync(
+            Document document,
+            SyntaxToken token,
+            CancellationToken cancellationToken)
+        {
+            // SuppressMessageAttribute docs 
+            // https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.suppressmessageattribute
+            var suppressMessageCheckIdArgument = token.GetAncestor<AttributeArgumentSyntax>() switch
```

Comment:
Usually, it should be enough to only walk up two or three parents (except the argument expression is deeply nested, which is unlikely for an attribute, but I have a test for that case anyway). I don't know whether it is problematic to walk up the entire tree here.

Reply:
@CyrusNajmabadi to answer this.